File: src/main/java/nl/insomnia247/nailbiter/eldrinkopubbot/state_machine/StateMachine.java
package nl.insomnia247.nailbiter.eldrinkopubbot.state_machine;
import nl.insomnia247.nailbiter.eldrinkopubbot.util.SecureString;
import java.util.HashMap;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.Comparator;
import java.util.Set;
import java.util.HashSet;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;
import org.apache.commons.lang3.tuple.ImmutablePair;
import java.lang.StringBuilder;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.LogManager;
import org.json.JSONArray;
import org.json.JSONObject;


/**
 * @author Alex Leontiev
 */
public class StateMachine<InputMessage,OutputMessage> implements Function<InputMessage,OutputMessage> {
    protected String _currentState = null;
    protected Set<String> _states = new HashSet<>();
    private static Logger _Log = LogManager.getLogger(StateMachine.class);
    protected Map<ImmutablePair<String,String>,List<ImmutablePair<Predicate<InputMessage>,Function<InputMessage,OutputMessage>>>> _transitions = new HashMap<>();
    public StateMachine(String state) {
        _currentState = state;
        _states.add(state);
    }
    public void inflateTransitionsFromJSON(Function<Object,Predicate<InputMessage>> conditionInflator, Function<Object,Function<InputMessage,OutputMessage>> actionInflator, String s) throws StateMachineException {
        JSONArray source = new JSONArray(s);
        for(int i = 0; i < source.length(); i++) {
            JSONArray a = source.getJSONArray(i);
            String start = a.isNull(0) ? null : a.getString(0)
                , end = a.isNull(1) ? null : a.getString(1);
            Predicate<InputMessage> pred = conditionInflator.apply(a.get(2));
            Function<InputMessage,OutputMessage> action = actionInflator.apply(a.get(3));

            _Log.info(SecureString.format("start: %s, end: %s",start,end));
            if(start!=null && end!=null) {
                this.addTransition(start,end,pred, action);
            } else if (start==null && end!=null) {
                for(String _start:this._states) {
                    this.addTransition(_start,end,pred,action);
                }
            } else {
                throw new StateMachineException("aed8ed251586aa75f5243c3a");
            }
        }
    }
    public StateMachine addTransition(String from, String to, 
            Predicate<InputMessage> transitionCondition, 
            Function<InputMessage,OutputMessage> transitionAction) {
        assert from != null;
        assert to != null;
        _states.add(to);
        _states.add(from);
        ImmutablePair<String,String> key = new ImmutablePair<>(from,to);
        if( !_transitions.containsKey(key) ) {
            _transitions.put(key,new ArrayList<ImmutablePair<Predicate<InputMessage>,Function<InputMessage,OutputMessage>>>());
        }
        _transitions
            .get(key)
            .add(new ImmutablePair<Predicate<InputMessage>, Function<InputMessage,OutputMessage>>(
                    transitionCondition,transitionAction
                    ));
        return this;
    }
    protected void _setState(String state) throws StateMachineException {
        if(!_states.contains(state)) {
            throw new StateMachineException(state);
        }
        _currentState = state;
        _onSetStateCallback(state);
    }
    @Override
    public OutputMessage apply(InputMessage im) {
        _Log.info(SecureString.format("apply: state: \"%s\"\nim: \"%s\"",_currentState,im));
        for(String to:_states) {
            List<ImmutablePair<Predicate<InputMessage>, Function<InputMessage,OutputMessage>>> pl
                = _transitions.get(new ImmutablePair<String,String>(_currentState,to));
            _Log.info(SecureString.format("checking %s -> %s [%s]",_currentState,to,pl));
            if( pl != null ) {
                for(ImmutablePair<Predicate<InputMessage>, Function<InputMessage,OutputMessage>> p:pl) {
                    if(p.left.test(im)) {
                        _Log.info(SecureString.format("active transition: %s -> %s",_currentState,to));
                        try {
                          _setState(to);
                        } catch (StateMachineException sme) {
                            return null;
                        }
                        OutputMessage om = p.right.apply(im);
                        _Log.info(SecureString.format("om: \"%s\"",om));
                        return om;
                    }
                }
            }
        }
        _didNotFoundSuitableTransition(im);
        return null;
    }
    protected void _didNotFoundSuitableTransition(InputMessage im) {
        _Log.info(SecureString.format("%s: did not found suitable transition. returning null",im));
    }
    protected void _onSetStateCallback(String state) {
        _Log.info(SecureString.format("state: %s\n",state));
    }
    @Override
    public String toString() {
        StringBuilder res = new StringBuilder();
        for(ImmutablePair<String,String> p : _transitions.keySet()) {
            res.append(SecureString.format("\"%s\" -> \"%s\"\n",p.left,p.right));
        }
        return res.toString();
    }
    public String toJsonString() {
        return toJsonString(new Function<Predicate<InputMessage>,Object>(){
            @Override
            public Object apply(Predicate<InputMessage> p) {
                return JSONObject.NULL;
            }
        }, new Function<Function<InputMessage,OutputMessage>,Object>() {
            @Override
            public Object apply(Function<InputMessage,OutputMessage> f){
                return JSONObject.NULL;
            }
        });
    }
    public String toJsonString(Function<Predicate<InputMessage>,Object> predicatePrinter, Function<Function<InputMessage,OutputMessage>,Object> transitionPrinter) {
        JSONObject res = new JSONObject();
        res.put("currentState",_currentState);

        JSONObject states = new JSONObject();
        for(String s:_states) {
            states.put(s,1);
        }
        res.put("states",states);

        JSONObject transitions = new JSONObject();
        for(ImmutablePair<String,String> p:_transitions.keySet()) {
            JSONArray val = new JSONArray();
            for(ImmutablePair<Predicate<InputMessage>,Function<InputMessage,OutputMessage>> pp: _transitions.get(p)) {
                JSONArray item = new JSONArray();
                item.put(predicatePrinter.apply(pp.left));
                item.put(transitionPrinter.apply(pp.right));
                val.put(item);
            }
            transitions.put(new JSONArray().put(p.left).put(p.right).toString(),val);
        }
        res.put("transitions",transitions);

        return res.toString();
    }
}

File: nl/insomnia247/nailbiter/eldrinkopubbot/state_machine/state_machine.py
"""===============================================================================

        FILE: /Users/nailbiter/Documents/forgithub/ElDrinkoPubBot/nl/insomnia247/nailbiter/eldrinkopubbot/state_machine/StateMachine.py

       USAGE: (not intended to be directly executed)

 DESCRIPTION: 

     OPTIONS: ---
REQUIREMENTS: ---
        BUGS: ---
       NOTES: ---
      AUTHOR: Alex Leontiev (alozz1991@gmail.com)
ORGANIZATION: 
     VERSION: ---
     CREATED: 2021-02-02T18:45:16.490735
    REVISION: ---

==============================================================================="""

import logging
from collections import namedtuple, OrderedDict
import json
import traceback
import uuid


class StateMachine:
    def __init__(self, state):
        self._current_state = state
        self._states = {state}
        self._transitions = []
        self._logger = logging.getLogger(self.__class__.__name__)

    def addTransition(self, from_, to, transition_condition, transition_action):
        self._states.add(from_)
        self._states.add(to)
        self._logger.info((from_, to, transition_condition))
        self._transitions.append(
            (from_, to, transition_condition, transition_action))

    def _setState(self, state):
        assert state in self._states, f"{state} not in {self._states}"
        self._current_state = state
        self._onSetStateCallback(state)

    def _onSetStateCallback(self, state):
        self._logger.info(f"set state: {state}")

    def __call__(self, input_message):
        try:
#            x = 1/0
#            self._logger.info(x)
            for k, state, transition_condition, transition_action in self._transitions:
                if k != self._current_state:
                    self._logger.info(f"{k}!={self._current_state}")
                    continue
                self._logger.info(f"checking condition {transition_condition}")
                if transition_condition(input_message):
                    self._logger.info("take")
                    self._setState(state)
                    res = transition_action(input_message)
                    self._logger.info(f"res: {res}")
                    return res
                else:
                    self._logger.info("pass")
        except Exception as exception:
#            self._logger.info("\n".join(traceback.format_stack()))
#            exc_fn = f"/tmp/{uuid.uuid4()}.log.txt"
#            with open(exc_fn,"w") as f:
#                traceback.print_tb(exception.__traceback__,file=f)
#            self._logger.info(f"exception data saved to {exc_fn}")
            self._exception_handler(exception, input_message)

        self._didNotFoundSuitableTransition(input_message)
        return None

    def _exception_handler(self, exception, input_message):
        self._logger.error(f"exception: {exception}")
        raise exception

    def _didNotFoundSuitableTransition(self, input_message):
        self._logger(
            f"{input_message}: did not found suitable transition -- returning None")

    def __str__(self):
        return "\n".join([f"{s} => {e}"for s, e, *_ in self._transitions])

#    def toJsonString(self, predicatePrinter=lambda _: None, transitionPrinter=lambda _: None):
#        return json.dumps({
#            "currentState": self._current_state,
#            "states": {s: 1 for s in self._states},
#            "transitions": {
#                json.dumps([s, e]): [[predicatePrinter(tc), transitionPrinter(ta)] for tc, ta in l]
#                for ((s, e), l)
#                in self._transitions.items()
#            },
#        })

    def inflateTransitionsFromJSON(self, conditionInflator, actionInflator, s):
        transitions = json.loads(s)
        for ss, es, *_ in transitions:
            self._states.update([x for x in [ss, es] if x is not None])
        self._logger.info(self._states)
        for ss, es, c, a in transitions:
            condition = conditionInflator(c)
            action = actionInflator(a, src_state=ss, dst_state=ss)
            if ss is not None and es is not None:
                self.addTransition(ss, es, condition, action)
            elif ss is None and es is not None:
                for s in self._states:
                    self.addTransition(s, es, condition, action)
            else:
                raise NotImplementedError(f"{ss,es}")
        self._logger.info([(ss, es, str(c))
                           for ss, es, c, a in self._transitions])
